go/vt/vtgate/vindexes/vschema.go

```diff
@@ -313,16 +337,54 @@ func (vschema *VSchema) Find(keyspace, tablename string) (table *Table, err erro
 	if !ok {
 		return nil, fmt.Errorf("keyspace %s not found in vschema", keyspace)
 	}
-	table = ks.Tables[tablename]
+	table := ks.Tables[tablename]
 	if table == nil {
 		if ks.Keyspace.Sharded {
-			return nil, fmt.Errorf("table %s not found", tablename)
+			return nil, nil
 		}
 		return &Table{Name: sqlparser.NewTableIdent(tablename), Keyspace: ks.Keyspace}, nil
 	}
 	return table, nil
 }
 
+// FindTableOrVindex finds a table or a Vindex by name using Find and FindVindex.
+func (vschema *VSchema) FindTableOrVindex(keyspace, name string) (*Table, Vindex, error) {
+	t, err := vschema.findTable(keyspace, name)
+	if err != nil {
+		return nil, nil, err
+	}
+	if t != nil {
+		return t, nil, nil
+	}
+	v, err := vschema.FindVindex(keyspace, name)
+	if err != nil {
+		return nil, nil, err
+	}
+	if v == nil {
+		return nil, nil, fmt.Errorf("table %s not found", name)
+	}
+	return nil, v, nil
+}
+
+// FindVindex finds a vindex by name. If a keyspace is specified, only vindexes
+// from that keyspace are searched. If no kesypace is specified, then a vindex
+// is returned only if its name is unique across all keyspaces. The function
+// returns an error only if the vindex name is ambiguous.
+func (vschema *VSchema) FindVindex(keyspace, name string) (Vindex, error) {
```

Comment:
In thinking about this now I disagree and think that this approach to name resolution isn't the best option.

One perspective to keep in mind is that sometimes (often?) the vindex name is actually irrelevant for uniqueness because it's a stateless vindex.

As a case in point, at slack we have the following vindex definitions in several of our keyspaces:

```
    "hash": {
      "type": "hash"
    },
    "md5": {
      "type": "unicode_loose_md5"
    },
```

So these vindexes are shared on several tables in the same keyspace and are repeated across keyspaces.

Another issue is that for clarity, whenever we have a lookup vindex, we use the same name as the backing table for the vindex itself, e.g.:

```
    "my_table_team_idx": {
      "type": "lookup_hash",
      "params": {
        "from": "team_id",
        "table": "my_table_team_idx",
        "to": "id"
      },
      "owner": "my_table"
    }
  },
```

This would obviously run into a name collision so there would be no way to actually query the vindex Map via this approach, since the backing table and the vindex are in the same keyspace and have the same name.

So, referring back to the design proposal https://github.com/youtube/vitess/blob/master/doc/VindexAsTable.md I'm wondering if it wouldn't make more sense to go with option 2 (a special `vindexes`) keyspace.

For all stateless vindexes like `hash`, it doesn't matter which version is picked, so we might as well use the first one. For lookup vindexes like `my_table_team_idx`, it is likely already unambiguous since it matches the name of the underlying lookup table.

Now, in the event that there are stateful vindexes which don't have unique names across keyspaces, we _could_ add similar routing logic as we do to resolve table name collisions and force the user to set a keyspace in the current session (i.e. by using `USE <keyspace` in the mysql protocol). This seems both unlikely to me and has a valid workaround to resolve the ambiguity.

It is of course a bit strange that `USE keyspace1; SELECT ... from vindexes.myvindex` would return something different from `USE keyspace2; SELECT ... from vindexes.myvindex`, but this whole approach is introducing synthetic pseudo-tables which don't map to underlying mysql constructs so I don't find this a show stopper.

All in all this feels to me like a more useful approach to naming than the current one.

Reply:
By design, vindexes are independent of the tables that use them; The `Map` function does not know which table is using it. Consequently, there can be no table-dependent behavior in there. This is the case for stateful vindexes also. So, it's safe to address a vindex as an independent entity.

The vindex design is a deviation from traditional RDBMS where tables encapsulate their indexes. In Vitess, vindexes have their own existence, which allows tables to share them.

To avoid name collisions, someone proposed using `_vdx` as suffix for all vindexes. I think it's a good idea, because it removes ambiguity while still rescuing you from inventing a new set of names. You could name your vindex `foo_vdx` and the underlying table `foo_idx`. I've been meaning to propose this as standard, and fix all our examples to follow it.

The alternative would be to introduce special syntax, which will reduce usability because tools like OR mappers can't do custom constructs.

The functionality you describe about qualifying vindexes by their keyspace is already implemented. I'm surprised I didn't add a test for it, which I'll do now.

Reply:
As an example of what I meant by "stateful vindexes", assume are two lookup vindexes, both named "my_lookup_vdx" in two different keyspaces, but each backed by a different table. I didn't mean that there was anything table dependent, just that the names don't necessarily convey the uniqueness of the vindex.

Proposing a naming convention like `_vdx` seems reasonable for best practices, and you're right that we could change the vschema to remove any name collisions.

But the part of this that makes me the most uneasy is the confusion to the user -- when I do `select keyspace_id from X where id = 123`, that could map to an in memory hash function, it could map to a lookup on an actual mysql table named X, or it could be a lookup to a different mysql table if X is a lookup vindex.

Presenting the vindexes as part of a reserved keyspace like `vindexes.X` is more in the spirit of things like the information_schema or performance_schema and as such doesn't suffer from the same confusion.

Reply:
I think the reserved keyspace approach goes against the design of the vschema: vindexes are defined inside a keyspace at the same level as tables. In this respect, it's easier to view vindexes as special tables.

If you look at it another way, tables have the same problem: there could be tables of the same name in different keyspaces. The same rules used to disambiguate tables gets uniformly applied to vindexes also.

We could even go one step further and prevent name collisions between vindexes and tables within a given keyspace. This will make vindexes resemble traditional tables even more.

Taking a step back, the information_schema style approach would have also been viable, probably even better. But we'd have to change the vschema model to match it.

Reply:
Thanks for the long explanation and sorry for my lag in replying.

Given all this I agree the current implementation is the best way to go even though I personally feel like the information_schema approach would have bene cleaner and easier to map to existing mysql concepts.